changeset:   990:1092b67b3cad
user:        mcimadamore
date:        Fri Apr 29 16:05:56 2011 +0100
summary:     7034495: Javac asserts on usage of wildcards in bounds

diff -r 9a847a77205d -r 1092b67b3cad src/share/classes/com/sun/tools/javac/code/Types.java
--- a/src/share/classes/com/sun/tools/javac/code/Types.java	Fri Apr 29 16:05:29 2011 +0100
+++ b/src/share/classes/com/sun/tools/javac/code/Types.java	Fri Apr 29 16:05:56 2011 +0100
@@ -412,6 +412,7 @@
                     return
                         s.tag == BOT || s.tag == CLASS ||
                         s.tag == ARRAY || s.tag == TYPEVAR;
+                case WILDCARD: //we shouldn't be here - avoids crash (see 7034495)
                 case NONE:
                     return false;
                 default:
diff -r 9a847a77205d -r 1092b67b3cad test/tools/javac/generics/wildcards/7034495/T7034495.java
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/test/tools/javac/generics/wildcards/7034495/T7034495.java	Fri Apr 29 16:05:56 2011 +0100
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2008, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug     7034495
+ * @summary Javac asserts on usage of wildcards in bounds
+ * @compile/fail/ref=T7034495.out -XDrawDiagnostics T7034495.java
+ */
+class T7034495 {
+
+    interface A<T> {
+        T foo();
+    }
+
+    interface B<T> {
+        T foo();
+    }
+
+    interface C<T extends A<?> & B<?>> { }
+
+}
diff -r 9a847a77205d -r 1092b67b3cad test/tools/javac/generics/wildcards/7034495/T7034495.out
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/test/tools/javac/generics/wildcards/7034495/T7034495.out	Fri Apr 29 16:05:56 2011 +0100
@@ -0,0 +1,2 @@
+T7034495.java:40:17: compiler.err.types.incompatible.diff.ret: T7034495.B<?>, T7034495.A<?>, foo()
+1 error

